### ISSUE TITLE
[doc request]execution order of one bulk operation

### ISSUE THREAD

**Issue by riverbuilding** [elastic/elasticsearch#50199]
according to this discuss:
https://discuss.elastic.co/t/order-of--bulk-request-operations/98124
You can rely on the fact that operations on the same document (same _index, _type and _id) will be in order. However you can't assume anything for documents that have different indices/types/ids.

according to this discuss:
https://discuss.elastic.co/t/is-the-execution-order-guaranteed-in-a-single-bulk-request/100412
If they are on the same shard, then yes.
If they aren't then we cannot guarantee that as far as I know.

according to this discuss:
https://discuss.elastic.co/t/elastic-search-bulk-request-api-is-the-order-of-document-writing-guaranteed/208055
Your bulk indexing request is received by a coordinating node. That node must then, document by document, determine the correct shard routing for each document, and then forward the documents to the nodes where each relevant shard is allocated.

Once the documents are distributed among the nodes, there can be no guarantees from node to node what order the indexing operations will be completed in, as the nodes are then working in parallel at whatever rate their available resources allow.

so can these facts be clearly put into doc:
1. (same _index, _type and _id) in same bulk will be in order
2. operations in same shard same bulk will be in order 
3. operations in same bulk but different shard has no order
4. operations in different bulk has no order

**Comment by markharwood** [comment]
Doesn't that logic simplify down to documenting point 2 - the one guarantee of sequencing?
Point 1 is a subset of point 2 and points 3 and 4 are ways to fail condition 2.

**Comment by ywelsch** [comment]
Even though 2 holds today, I think we should leave that unspecified, as it limits how bulk requests can be processed at the shard level. Think e.g. if we wanted to introduce more concurrency at the bulk shard request level.

**Comment by markharwood** [comment]
Thanks for the suggestion but I will close this issue as "won't fix" for the reason given above.
